pkg/sentry/fs/proc/net.go

```diff
@@ -194,6 +195,199 @@ func (n *netDev) ReadSeqFileData(ctx context.Context, h seqfile.SeqHandle) ([]se
 	return data, 0
 }
 
+// netSnmp implements seqfile.SeqSource for /proc/net/snmp.
+//
+// +stateify savable
+type netSnmp struct {
+	s inet.Stack
+}
+
+// NeedsUpdate implements seqfile.SeqSource.NeedsUpdate.
+func (n *netSnmp) NeedsUpdate(generation int64) bool {
+	return true
+}
+
+type snmpLine struct {
+	prefix string
+	header string
+}
+
+var snmp = []snmpLine{
+	snmpLine{
+		prefix: "Ip",
+		header: "Forwarding DefaultTTL InReceives InHdrErrors InAddrErrors ForwDatagrams InUnknownProtos InDiscards InDelivers OutRequests OutDiscards OutNoRoutes ReasmTimeout ReasmReqds ReasmOKs ReasmFails FragOKs FragFails FragCreates",
+	},
+	snmpLine{
+		prefix: "Icmp",
+		header: "InMsgs InErrors InCsumErrors InDestUnreachs InTimeExcds InParmProbs InSrcQuenchs InRedirects InEchos InEchoReps InTimestamps InTimestampReps InAddrMasks InAddrMaskReps OutMsgs OutErrors OutDestUnreachs OutTimeExcds OutParmProbs OutSrcQuenchs OutRedirects OutEchos OutEchoReps OutTimestamps OutTimestampReps OutAddrMasks OutAddrMaskReps",
+	},
+	snmpLine{
+		prefix: "IcmpMsg",
+	},
+	snmpLine{
+		prefix: "Tcp",
+		header: "RtoAlgorithm RtoMin RtoMax MaxConn ActiveOpens PassiveOpens AttemptFails EstabResets CurrEstab InSegs OutSegs RetransSegs InErrs OutRsts InCsumErrors",
+	},
+	snmpLine{
+		prefix: "Udp",
+		header: "InDatagrams NoPorts InErrors OutDatagrams RcvbufErrors SndbufErrors InCsumErrors IgnoredMulti",
+	},
+	snmpLine{
+		prefix: "UdpLite",
+		header: "InDatagrams NoPorts InErrors OutDatagrams RcvbufErrors SndbufErrors InCsumErrors IgnoredMulti",
+	},
+}
+
+func sprintSlice(s []uint64) string {
+	if len(s) == 0 {
+		return ""
+	}
+	return strings.Trim(fmt.Sprint(s), "[]")
+}
+
+// ReadSeqFileData implements seqfile.SeqSource.ReadSeqFileData. See Linux's
+// net/core/net-procfs.c:dev_seq_show.
+func (n *netSnmp) ReadSeqFileData(ctx context.Context, h seqfile.SeqHandle) ([]seqfile.SeqData, int64) {
+	if h != nil {
+		return nil, 0
+	}
+
+	var (
+		snmpIp      inet.StatSnmpIp
+		snmpIcmp    inet.StatSnmpIcmp
+		snmpTcp     inet.StatSnmpTcp
+		snmpUdp     inet.StatSnmpUdp
+		snmpUdpLite inet.StatSnmpUdpLite
+	)
+	contents := make([]string, len(snmp)*2)
```

Comment:
igudger: This is a lot of duplicated code. Consider using a loop.

contents := make([]string, len(snmp)*2, 0)

types := []interface{}{
	&inet.StatSnmpIp{},
	&inet.StatSnmpIcmp{},
	nil,
	&inet.StatSnmpIcmp{},
	&inet.StatSnmpIcmp{},
	&inet.StatSnmpIcmp{},
}

for i, stat := range types {
	if stat == nil {
		contents = append(contents, "", "")
		continue
	}
	line = snmp[i]
	if err := n.s.Statistics(stat, line.prefix); err != nil {
		log.Warningf("Failed to retrieve %s of /proc/net/snmp: %v", line.prefix, err)
	}
	contents = append(
		contents,
		fmt.Sprintf("%s: %s\n", line.prefix, line.header),
		fmt.Sprintf("%s: %s\n", line.prefix, sprintSlice(toSlice(stat))),
	)
}

toSlice: https://play.golang.org/p/mmIgCCOoaAb

https://gvisor-review.git.corp.google.com/c/gvisor/+/18283/6/pkg/sentry/fs/proc/net.go#301

Reply:
Done.